### ISSUE TITLE
Feature: Packaging of Node

### ISSUE THREAD

**Issue by vsoch** [apptainer/singularity#782]
This is a summary of a feature that (is going to be awesome) and have many use cases for 2.4. It's the idea that someone can package the node that they are working on, and doesn't have to know a thing about containers to make them.

I think this could be integrated into Singularity as a bootstrap module, but for now I've implemented something to test [here](https://github.com/vsoch/singularity-python/tree/development/examples/create/package_node). Specifically, it looks something like this:

## On Node
```
from singularity.package import package_node
package = package_node()
 
package
'/tmp/tmp06ww_8_n/vanessa-testing-self-cloning.tgz'
```

optionally, the user can select a different root (default is `/`) or name for the image (the default name is the name of the machine)

## On Packaging Node

```
from singularity.package import unpack_node
image = unpack_node(image_path=package,size=9000)
```

The current issues with the above have to do with the metadata folder (eg, there is no action for the container, or anything else for that matter) but if we integrate this into bootstrap, that goes away.

## Workflow
This supports  a slightly different framework than is supported on most clusters - one that is more akin to requesting instances on demand on Google Cloud (and likely some larger resources do have with something like OpenStack). Under this design, a user is given a node on demand, disconnected from the cluster, and allowed to do whatever they please!

 - history is saved as a quasi "deffile" %post section, to record exactly what they did
- full permission is given to install, clone, etc.
 
When the user has finished, a trigger is run to package the image, and sent somewhere (a different machine with enough space) to package it into an image proper (the unpack_image function above, which would be supported through singularity hopefully to create a completely working image).

## Features
For an actual imlpementation, we would want to have the following:
 - security checks for the image before it's done
 - testing in the same way
 - capturing metadata, akin to what singularity hub does (which is pretty crappy, we can do a lot better)!
 
## TLDR

 1. user logs into node
 2. user does work, perhaps it's recorded in history (to save and mimic as some kind of deffile)
 3. when user is finished, some kind of trigger is done (or running a command for an admin to do so) to package the node
 4. the node is packaged, moved somewhere where it can be tested, measured, and turned into an image.
 5. the image is made available for the user.

**Comment by truatpasteurdotfr** [comment]
- history cleanup (no trace of credentials like "echo changemenow| sudo passwd --stdin root")
- user controlled whitelist/blacklist of what should be kept
- remove any file specific to the machine and not required on the container (same files as for cloning instances) e.g /etc/ssh/*host*,  cleanup /etc/hosts, /etc/sysconfig/network-scripts/ifcfg-*, but not ifcfg-lo, /etc/udev/rules.d/*persistent* ) /var/log/{secure*,messages*,...}
- yum clean all and the apt-get equivalent
- SElinux or ACLs transfer? tar --selinux --xattrs
- sparse file handling
- remove any non required users via userdel?

maybe reuse|have a peak at cloud-init ?

**Comment by truatpasteurdotfr** [comment]
maybe http://libguestfs.org/virt-sysprep.1.html#operations would be a better start

**Comment by truatpasteurdotfr** [comment]
or ping @gmkurtzer  for the warewulf vnfs tidbits used on the warewulf framework ;P

**Comment by dtrudg** [comment]
This is something that is outside of the scope of a Singularity feature, but remains interesting as a separate concept. It is likely that plans to allow plugin hooks for the build engine in the next couple of versions of Singularity will permit it to be tackled by anyone who is interested.
